File: ruby/word-count/word_count.rb
module BookKeeping
  VERSION = 1
end

class Phrase
  def initialize(string)
    @string = string
  end

  def word_count
    @string.split(/[\s\W&&[^']]/).each_with_object({}) do |word, acc|
      next if word.empty?
      word = word.downcase.gsub(/\A'|'\Z/, '')
      acc[word] ||= 0
      acc[word] += 1
    end
  end
end

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map
  def count(sentence) do
    sentence
    |> String.downcase()
    |> String.split(~r/[^\p{L}\d\-]/u, trim: true)
    |> Enum.reduce(%{}, &update_count/2)
  end

  defp update_count(word, acc) do
    Map.update(acc, word, 1, &(&1 + 1))
  end
end
